### ISSUE TITLE
opcode || doesn't work

### ISSUE THREAD

**Issue by kesonIT** [containous/yaegi#474]
The following program print wrong result

```go
c := false
println("c || !c = ", c || !c)
println("!c || c = ", !c || c)
```

Expected result:
```
c || !c =  true
!c || c =  true
```

Got:
```
c || !c =  false
!c || c =  false
```

**Comment by ldez** [comment]
duplicate of #471,  #465

**Comment by kesonIT** [comment]
hi @ldez ,  it is another problem, please take a look at the details

**Comment by kesonIT** [comment]
@ldez 471 doesn't fix this problem

**Comment by mvertes** [comment]
The bugs are real legitimate. They concern potentially all operations that return boolean values and that can be used in branch operations.

The bug comes in fact from a wrong optimization, where the result of a boolean operation (assign, compare, etc...) is not stored for future test, but used directly for branching. Unfortunately, that doesn't work with current impelementation of logical or / and operations, which rely explicitely on a stored result to decide.

I fixed the problem for #471 for `isNil()` and `isNotNil()` in #473, but I realize that I have to generalize the fix to all boolean operations.

I propose to reopen #471 for that and close other bugs in favor of it. What do you think ?
